### ISSUE TITLE
Clarify the responses that should be given when a request is missing arguments

### ISSUE THREAD

**Issue by turt2live** [matrix-org/matrix-doc#1437]
Dendrite seems to prefer a 400 `M_MISSING_ARGUMENTS` while Synapse goes for the 500 `M_UNKNOWN` route. A consistent error would be nice not only amoung the 2 projects, but in the spec as well.

Applies to all APIs, not just c2s.

**Comment by richvdh** [comment]
It is very obviously a bug to return a 500 error in response to missing or unexpected parameters.

**Comment by turt2live** [comment]
Yes, but the problem is that it's not in the spec, so it can't entirely be declared a bug despite being obvious.

**Comment by richvdh** [comment]
maybe it should be in the spec. Either way, it's http error code lesson 101:

> 1xx: hold on
2xx: here you go
3xx: go away
4xx: you fucked up
5xx: I fucked up

I still say it's an obvious bug in synapse which should be fixed.

As for M_UNKNOWN: I'd love it if the spec could say something like "Servers SHOULD only return M_UNKNOWN when no other error code is appropriate. Clients SHOULD NOT rely on it being returned under any particular set of circumstances."

**Comment by turt2live** [comment]
out of band discussion yielded synapse being buggy in this area. Leaving this open as a general "make it  clearer" thing for the spec.
